torch/csrc/distributed/rpc/PythonRpcHandler.cpp

```diff
@@ -0,0 +1,89 @@
+#include <torch/csrc/distributed/rpc/PythonRpcHandler.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+PyObject* PythonRpcHandler::module_ = nullptr;
+PyObject* PythonRpcHandler::runUDFFunction_ = nullptr;
+PyObject* PythonRpcHandler::loadResultFunction_ = nullptr;
+
+void PythonRpcHandler::init() {
+  Py_Initialize();
```

Comment:
I think we need to discuss more about this line.

It looks to me that our user starts with launching a Python interpreter (a C program) to interpret a piece of Python code. This Python interpreter, loads a pybind11 module, `torch.distributed`, that bundles several C++ classes.
We use this special binding module, backed by C++ classes, to instantiate a `RpcAgent` C++ instance, and assigned it to a variable `torch.distributed.rpc._agent`.
While, in the constructor of `RpcAgent`, `PythonRpcHandler::init()` is called and creates a **second** Python interpreter ?? (If I understand the [doc](https://docs.python.org/3/c-api/init.html#c.Py_Initialize) correctly)
This **second** Python interpreter imports the module, `torch.distributed.internal_rpc_utils`, and use the utilities in it. 
Although it can successfully run the function sent by the client and return correct results back to the client.
It's impossible for the client to change the global state in the server, which is hold by the **first** Python interpreter.

---
There is a requirement in our RRef prototype implementation that we need to allow an RPC client (an RRef user), to dictate the RPC server (RRef owner) to change it's global state. 
The state is stored in a module's global variable, which lives in the original Python Interpreter.

For example, 
1) A client could ask the server to fetch an RRef instance that matches the `ref_id` in the message, from the server's global RRef registry.
2) A client could send termination signal message to a server, the server, on receiving the signal, should set it's status to `termination_ongoing=True`, so that it will delete the remote references it holds and notifies the owners about reducing ref counts before it goes away.
---
I read this doc, https://docs.python.org/3/extending/index.html#extending-index

It looks to me there are 2 ways for C/C++ to interact with Python.

1) > Write modules in C or C++ to extend the Python interpreter with new modules.
2) > Sometimes, rather than creating an extension that runs inside the Python interpreter as the main application, it is desirable to instead embed the CPython runtime inside a larger application.

Looks to me `Py_Initialize()` belongs to the interface for the seconds case.

Reply:
I thought we only call Py_Initialize() once to import interpreter, as init() only is called once...? although we can enforce once as well, but I do not understand why it is called twice?

also could you please share what kind of error you encountered for the integration? we can sync up offline

Reply:
I thought Py_Initialize() will only be called once, as init() is called only once...? although I can enforce it to be called only once explicitly as well, but I do not understand how it  could be called twice to import interpreter? we can sync up offline about this

Reply:
@zhaojuanmao 
NVM. There is a call to create a new Interpreter.
https://docs.python.org/3/c-api/init.html#c.Py_NewInterpreter

So I think `Py_Initialize` ensures singleton and initialize only once.